### ISSUE TITLE
Out of memory, zooming, ZoomPanListener

### ISSUE THREAD

**Issue by chrisbrossard** [moagrius/TileView#301]
Thanks for this wonderful library.  I have put together several mapping apps which display tiles downloaded from servers.  They have a fairly large number of zoom levels (up to 20).  I have also created a couple of buttons for "zoom in" and "zoom out" which double and halve the scale respectively. 
I have noticed that if I set the initial zoom level (scale) so that the map is "zoomed out", the library crashes with an out of memory error.  The exact zoom level at which this happens depends on the device.  I have worked around this by initializing the library so that the map starts out quite "zoomed in".
I have also noticed that if I start the map "zoomed in", and zoom out using my zoom buttons, no out of memory error occurs.
Also, the "zoom wraparound", in which the scale of the map goes from its minimum to maximum value when the map is double tapped also causes an out of memory error in the library.  I have intercepted this and blocked it to prevent the crash.
Also, I notice that occasionally when the zoom out button is tapped, my "getBitmap" callback is not called by the library.  If I then pan the map, my "getBitmap" callback is called and the map fills correctly.
Also, I have noticed that the current detail level is not set correctly when the "onZoomEnd" callback is called in the ZoomPanListener .
Finally, I have noticed from analyzing crashes, and just wanted to confirm, that the "getBitmap" callback is called from an AsyncTask, so there is no need for me to start up my own AsyncTask in getBitmap() when downloading a tile.
Again, thanks for the library.

**Comment by peterLaurence** [comment]
> Also, I notice that occasionally when the zoom out button is tapped, my "getBitmap" callback is not called by the library. If I then pan the map, my "getBitmap" callback is called and the map fills correctly.
Also, I have noticed that the current detail level is not set correctly when the "onZoomEnd" callback is called in the ZoomPanListener .

@moagrius Reminds me a discussion we had about one of my first PR. I talked about the potential OOM when zooming out on big maps (i explained why the `DetailLevel` lock while zooming can cause that). It would be nice to fix that while keeping the lock.

**Comment by moagrius** [comment]
Not related to TileView, but I run into OOME all the time still, especially when supporting old devices - Android Devices have a limited heap, and bitmap manipulations are one way to run out in a hurry.  On one hand, I'm inclined to say that if you're getting OOME when zooming out, your probably just need more small detail levels, but on the other hand, if you've got something in mind @peterLaurence, I'd obviously be very happy to make the library as good as possible.

> I have noticed that if I set the initial zoom level (scale) so that the map is "zoomed out", the library crashes with an out of memory error. The exact zoom level at which this happens depends on the device. I have worked around this by initializing the library so that the map starts out quite "zoomed in".
I have also noticed that if I start the map "zoomed in", and zoom out using my zoom buttons, no out of memory error occurs.

The only thing I can really say here is to try to use detail levels and scale as wisely as you can.  Remember that if we scale _down_ bitmaps, they still take just as much memory, so if you have a tile set that can conceivably be 25% or smaller of actual size, that means you're going to have 400% plus pixels being drawn that the screen can support.  Small detail levels are more important than large ones...

> Also, the "zoom wraparound", in which the scale of the map goes from its minimum to maximum value when the map is double tapped also causes an out of memory error in the library. I have intercepted this and blocked it to prevent the crash.

Same advice above applies.  Something to be aware of is that when the TileView changes detail levels, it renders the most current one over the last one - this is the only time there are more than 1 "layer" of tiles, and it happens because otherwise the UX would be weird as the old tiles just vanished and we waited for the new set to render.  I suppose ideally, we'd invalidate regions as new tiles were rendered, but that's a _heavy_ lift and not in the cards for me at the moment (although always welcome quality PRs)...

> Also, I have noticed that the current detail level is not set correctly when the "onZoomEnd" callback is called in the ZoomPanListener .

This one sounds like a bug...  @peterLaurence can you confirm this off the top of your head?  if not, no worries, I can check it out.

> Finally, I have noticed from analyzing crashes, and just wanted to confirm, that the "getBitmap" callback is called from an AsyncTask, so there is no need for me to start up my own AsyncTask in getBitmap() when downloading a tile.

We don't use AsyncTasks, but all decoding does happen in background threads.

Thanks for posting - I hope the above doesn't sound dismissive - honestly trying to avoid OOME is a full time job.  Feel free to post back with ideas, and if you have some idea on how to fix the issues you're describing, we'd be happy to take a look.

**Comment by peterLaurence** [comment]
I found my comment about a related issue. It explains how to reproduce it and talks about the related code. Even though this was valid for TileView v2.0, there is a good chance that it still applies (i don't think that `DetailLevel` change logic was touched) :

> [to reproduce]  you need a zoom that will cause a render of a lot of tiles. Double-tap from max scale is perfect. You need a map with enough DetailLevel. The demo may be enough. To emphasize this, set TileView.setShouldScaleToFit( false ) and have a minimum scale equal or below the minimum DetailLevel's scale to be be able to see every levels (and trigger the issue). So from max scale you double-tap. What happens next? Because of the lock, the DetailLevel don't change until the zoom ends. But before it ends, numerous tiles from that level will be rendered and scaled because during the zoom, TileView.onScrollChanged is being called (which indirectly calls requestRender). As a side note, that's were we have a likelyhood of OOM. When the zoom ends, it removes the lock, updates the scale of the DetailLevelManager, which in turn calls DetailLevelManager.update() and then triggers a DetailLevel change. Given the implementation of a DetailLevel change in TileView :
```
@Override
public void onDetailLevelChanged( DetailLevel detailLevel ) {
   requestRender();
   mTileCanvasViewGroup.updateTileSet( detailLevel );
}
```
> the updateTileSet calls a cancelRender. So it attempts to stop the ongoing task mentioned before (which may have not finished yet because of the number of the tiles to render). At this point we can say : "that's ok, after that a requestRender is called and everything will be fine". BUT, it turns out, that request may just have a DetailLevel.computeCurrentState() returning false, so the requestRender does nothing. So why computeCurrentState may return false? While zooming, a new requestRender may call computeCurrentState at a point were the current DetailLevel is equal to the "destination" one (idk how to say this :). So, in DetailLevel, the attribute mLastStateSnapshot already has its end value, and so as the other parameters rowStart, columnStart, ...
And there we have it : the requestRender that does nothing! :)

> To avoid this, i distinguished the DetaiLevel thats has been requested to render from the one that has really been rendered. In the case exposed above, the requested DetailLevel is the final one (min scale) while the rendered one is most likely the one with max scale. So at the end of the zoom, even if the computeCurrentState sees no difference, the additional check i added avoids the straight return and the min scale DetailLevel is rendered.

The final fix mentioned at the end was never integrated.
To sum up, i think that the OOM issue described in the OP may be due to the `DetailLevel` lock plus the scale change. And the `DetailLevel` not set correctly is discussed in the second part of my comment.

**Comment by peterLaurence** [comment]
So about the OOM,

@chrisbrossard could you try to disable the `DetailLevel` lock (comment the `mDetailLevelManager.lockDetailLevel();` and `mDetailLevelManager.unlockDetailLevel();` in `onZoomBegin` and `onZoomEnd` respectively). Make your testings and report back if you have still the OOM. If you haven't, it will give a good hint on what's happening. And we will discuss whether this needs a library fix or not.

**Comment by chrisbrossard** [comment]
I added the calls to unlockDetailLevel() in onZoomBegin and onZoomEnd and the OOM problem has disappeared.

**Comment by peterLaurence** [comment]
@chrisbrossard it's another (yet strange) way to do it.
Ok so that's what i expected. Can you tell us the scales of your detail levels ?

**Comment by chrisbrossard** [comment]
myTileView.addDetailLevel(1.00000000000000f, "20", TILE_WIDTH_PIXELS, TILE_HEIGHT_PIXELS);
        myTileView.addDetailLevel(0.50000000000000f, "19", TILE_WIDTH_PIXELS, TILE_HEIGHT_PIXELS);
        myTileView.addDetailLevel(0.25000000000000f, "18", TILE_WIDTH_PIXELS, TILE_HEIGHT_PIXELS);
        myTileView.addDetailLevel(0.12500000000000f, "17", TILE_WIDTH_PIXELS, TILE_HEIGHT_PIXELS);
        myTileView.addDetailLevel(0.06250000000000f, "16", TILE_WIDTH_PIXELS, TILE_HEIGHT_PIXELS);
        myTileView.addDetailLevel(0.03125000000000f, "15", TILE_WIDTH_PIXELS, TILE_HEIGHT_PIXELS);
        myTileView.addDetailLevel(0.01562500000000f, "14", TILE_WIDTH_PIXELS, TILE_HEIGHT_PIXELS);
        myTileView.addDetailLevel(0.00781250000000f, "13", TILE_WIDTH_PIXELS, TILE_HEIGHT_PIXELS);
        myTileView.addDetailLevel(0.00390625000000f, "12", TILE_WIDTH_PIXELS, TILE_HEIGHT_PIXELS);
        myTileView.addDetailLevel(0.00195312500000f, "11", TILE_WIDTH_PIXELS, TILE_HEIGHT_PIXELS);
        myTileView.addDetailLevel(0.00097656250000f, "10", TILE_WIDTH_PIXELS, TILE_HEIGHT_PIXELS);
        myTileView.addDetailLevel(0.00048828125000f, "9", TILE_WIDTH_PIXELS, TILE_HEIGHT_PIXELS);
        myTileView.addDetailLevel(0.00024414062500f, "8", TILE_WIDTH_PIXELS, TILE_HEIGHT_PIXELS);
        myTileView.addDetailLevel(0.00012207031250f, "7", TILE_WIDTH_PIXELS, TILE_HEIGHT_PIXELS);
        myTileView.addDetailLevel(0.00006103515625f, "6", TILE_WIDTH_PIXELS, TILE_HEIGHT_PIXELS);
        myTileView.addDetailLevel(0.00003051757813f, "5", TILE_WIDTH_PIXELS, TILE_HEIGHT_PIXELS);
        myTileView.addDetailLevel(0.00001525878907f, "4", TILE_WIDTH_PIXELS, TILE_HEIGHT_PIXELS);
        myTileView.addDetailLevel(0.00000762939454f, "3", TILE_WIDTH_PIXELS, TILE_HEIGHT_PIXELS);
        myTileView.addDetailLevel(0.00000381469727f, "2", TILE_WIDTH_PIXELS, TILE_HEIGHT_PIXELS);
        myTileView.addDetailLevel(0.00000190734864f, "1", TILE_WIDTH_PIXELS, TILE_HEIGHT_PIXELS);

**Comment by peterLaurence** [comment]
Apparently, you have created every detail levels for scales between 1 and 1/2^19.

@moagrius his case illustrates the OOM risk i was previously talking about. I personally never encountered it, probably because i have much less detail levels (around 7) in my app. So my thought is that we could add an option like `setShouldUpdateDetailLevelWhileZooming`. The default would be `false`. When true, we would have the UX updated in a way you don't like, i know, but at least the `TileView` won't try to load too much tiles of a detail level at a "too low" scale compared to the current scale.
That leads to another way to fix that issue : maybe introduce a threshold to the number of tiles to render for each detail level (or the whole detail levels). That threshold could be optimized by the developer for the target device after tests.
I prefer the first option, but perhaps you have a better idea.

**Comment by moagrius** [comment]
@peterLaurence the first option sounds reasonable, at least as a bandaid until we can really consider the best way to handle it...  ideally, we'd be writing to a single Canvas and just invalidating where appropriate, instead of stacked layers, but that's a _big_ ask, and not one i have time for now...

do you have the bandwidth to handle this one (your first suggestion, `setShouldUpdateDetailLevelWhileZooming`) or do you want me to?

**Comment by peterLaurence** [comment]
@moagrius i will do it soon.

**Comment by moagrius** [comment]
thank you sir

**Comment by peterLaurence** [comment]
@chrisbrossard TileView 2.1.6 is out and you can you can use `TileView.setShouldUpdateDetailLevelWhileZooming( true )` to avoid OOM in your particular case. With that option used, could you make some tests and report back if you still have some issues (like the "getBitmap" callback not being called..)?

**Comment by chrisbrossard** [comment]
@peterLaurence I removed the calls to unlockDetailLevel() and changed the module Gradle file to include version 2.1.6.  The results are that I can now start the library with the map "zoomed out" and there is no OOM error.  The zoom wraparound also now works correctly.  I also noticed that the tiles now download very quickly.  The problem of getBitmap not being called sometimes when zooming out still persists.  I added more Log statements to getBitmap() and discovered that for the zooming out test for this particular app, the library asks for tiles 3-0-5 and 3-5-7 (level-column-row) which, as far as I can tell, are being returned to the library.  The library then asks for no more tiles.  Also, those two tiles are being downloaded simultaneously.  That is, before the first tile is returned to the library, the library asks getBitmap for the second tile.  I'm not sure if this is okay or if I'm supposed to block the second request with a semaphore until the first tile is returned to the library.  Just to be absolutely clear though, I'm not currently using any semaphores.  I'm just wondering if I should be, or if in general it is okay to download multiple tiles simultaneously, which I guess means having multiple HTTP connections going simultaneously.

**Comment by peterLaurence** [comment]
At he moment, we need to understand what's really happening. I have an idea of the root cause but i need to be sure. So, to sum up :

- on a zoom out, some tiles are not rendered at the destination scale
- a pan can fix that and the library renders the missing tiles (to confirm that, don't use the wrap around zoom because you will end up at the minimum scale where you might be able to pan).

Correct me if this is wrong.

**Comment by chrisbrossard** [comment]
What you have written is correct.

**Comment by peterLaurence** [comment]
@chrisbrossard As i can't reproduce the issue on my side, i'd like to make a new commit in my personal fork of TileView in a new branch. Will you be able to compile and test it?
It's simple, you can generate the aar by compiling the project then import that aar but temporarily remove the current 2.1.6 in gradle file.
Please confirm this is ok for you.

**Comment by chrisbrossard** [comment]
If I understand correctly, you want me to download the source for your test version of the library, then use Android Studio to create an Android Archive Library, then manually import that into my project after removing the standard 2.1.6 version of the library from my project.  Yes, I think I can figure out how to do that.

**Comment by peterLaurence** [comment]
Ok good. Here is the [link](https://github.com/peterLaurence/TileView/tree/fix%23252) to my TileView fork, branch "fix#252".
You can try this. I hope it fixes your issue.

**Comment by chrisbrossard** [comment]
Yes that has fixed the zoom out issue.

**Comment by peterLaurence** [comment]
@chrisbrossard thanks for testing and feedback.

@moagrius new PR incoming!

**Comment by moagrius** [comment]
nice work

**Comment by peterLaurence** [comment]
@chrisbrossard TileView 2.1.7 will be published soon. This will include the fix you got from my fork.

**Comment by bb-juliogarcia** [comment]
Hi, although I am not experiencing any OOM errors on version 2.1.7, I am still experiencing missing tiles when zooming out quickly to the minimum scale. I've tried using an object pool, and have also tried using `tileView.setShouldUpdateDetailLevelWhileZooming( true ) ` but still have missing tiles. Any ideas on why this could still be happening? @moagrius @peterLaurence

**Comment by moagrius** [comment]
You're probably running out of memory and catching it silently - 
 Are you using the built in asset decoder?  That catches OOME https://github.com/moagrius/TileView/blob/master/tileview/src/main/java/com/qozix/tileview/graphics/BitmapProviderAssets.java#L42 - either 
 log that out or printStackTrace or use your own implementation.  If that is the case, try to get the file size of your tiles images down as much as possible - there are utilities online that do a great job of that.  Also use more detail levels so there's as few pixels to draw as possible.  Imagine if your smallest detail level has to be scaled down a lot, the amount of pixel data continues to grow.

HTH, GL

**Comment by bb-juliogarcia** [comment]
@moagrius I am actually using a custom decoder that reads from a FileInputStream that is very similar to your built in one. In fact, after making modifications to try and catch the OOME, I see that I can never hit the breakpoint in the catch block so I am not sure what is going on.

For TileView configuration, the relevant lines are 

`      tileView.setShouldUpdateDetailLevelWhileZooming(true);
        tileView.setShouldRenderWhilePanning(false);
        tileView.setBitmapProvider(new TileBitmapProvider());`

And for the custom TileBitmapProvider here is what I have

```
public Bitmap getBitmap(Tile tile, Context context) {
        Object data = tile.getData();
        String unformattedFileName = (String) data;
        String formattedFileName = String.format(unformattedFileName, tile.getColumn(), tile.getRow());
        Bitmap bitmap = null;
        try {
            InputStream stream = new FileInputStream(formattedFileName);
            bitmap = BitmapFactory.decodeStream(stream, null, OPTIONS);
        } catch(  OutOfMemoryError | Exception e ) {
            e.printStackTrace(); /* Breakpoint here */
        } finally {
            return bitmap;
        }
}
```

For some reason, I can't get it to hit the OOME breakpoint, yet I still see blank tiles at the lowest detail level (zoomed out on a 3x3 tile grid at 512px by 512px). Each tile is approximately 25-40kb, and I can confirm the tiles exist as I can see them when I zoom in and out at a normal pace. Just not so when I do it quickly. Thanks for your help!

**Comment by moagrius** [comment]
i'd probably just log out file exists or file size - not sure why it won't hit that breakpoint - try logs? - also you don't need the OOME in your catch tho (just `Exception e` is enough)

gl!

**Comment by peterLaurence** [comment]
The issue @bb-juliogarcia described also happens to me sometimes. But it's rare and just panning/zooming in a little bit makes the missing tiles appear again. 
I never took the time to investigate further, mostly because the problem is not easy to reproduce with my devices (although it should trigger more often on emulator).
I will have look at this.

**Comment by peterLaurence** [comment]
Investigating on this is not easy, but i think i have narrowed the issue. 
My step to reproduce is to launch the demo with the large image, and allow the scale loop by commenting the line `tileView.setShouldLoopScale( false );`
I double click the image to circle through all scales. Doing that ~5 times i finally get :
![capture du 2017-11-05 00-49-41](https://user-images.githubusercontent.com/15638794/32410570-fd5d1d82-c1c3-11e7-9083-d5947dab86f3.png)

I first looked at image decoding issues but there were no error at all. So the issue is most probably in the tile rendering logic in `TileCanvasViewGroup`. 
Interesting fact regarding the following method :
```
private void drawTiles( Canvas canvas ) {
    if( mPreviouslyDrawnTiles.size() > 0 ) {
      // Case 1
      drawTilesConsideringPreviouslyDrawnLevel( canvas );
    } else {
      // Case 2
      drawTilesWithoutConsideringPreviouslyDrawnLevel( canvas );
    }
  }
```
When there is no issue when going back to the first detail level, the case 2 is called after the case 1 were called multiple times. This is perfectly repeatable, until the issue triggers and no case 2 is called. Tracking the root cause for this led to the fact that in this particular case `mTilesInCurrentViewport` is empty.

Long story short, by slightly modifying the `beginRenderTask` method to :
```
private void beginRenderTask() {
    // if visible columns and rows are same as previously computed, fast-fail
    boolean changed = mDetailLevelToRender.computeCurrentState();
    if( !changed && mDetailLevelToRender.equals( mLastRenderedDetailLevel) && mTilesInCurrentViewport.size() > 0) {
      System.out.println("fail");
      return;
    }
// rest of the code
```
**I could not reproduce the issue after that change.**
So i have a sort of fix. But i don't yet understand why in that case `mTilesInCurrentViewport` is empty.

**Comment by peterLaurence** [comment]
After some more research, my understanding of why `mTilesInCurrentViewport` is empty is if detail level changes happen fast enough, we encounter the following situation :

A high amount of request render are kicked in `TileCanvasViewGroup.requestRender()` which is based on a throttle logic. Those requests which usually pass through this enable the launch of new tile decoding with the `TileRenderPoolExecutor`, in the `beginRenderTask`. So since no new task are sent to the executor, the rendering mechanism is broken.
The modification I suggested in my last post enables the last request render to not fast-fail and trigger the last detail level render.

@moagrius what do you think?

**Comment by peterLaurence** [comment]
Sorry to bother you again but i found another fix, which consists of removing the fast-fail test and thus all boilerplate code related to `mLastRenderedDetailLevel`.
The fast-fail statement is there to avoid unnecessary calls to the rendering engine or even go into an infinite loop. Turns out, removing it cause neither of these problems. Simpler is better...

As for the OOM issue reported in the OP, this has been fixed by the removing of detail level lock. This fixes the tiles missing issue.

**Comment by bb-juliogarcia** [comment]
@peterLaurence Thanks for taking the time to look into this. Your efforts are greatly appreciated!

**Comment by moagrius** [comment]
hey @peterLaurence sorry for the delay - never a bother - do you want to create a PR?

**Comment by peterLaurence** [comment]
@moagrius Sure! Will do.
